### DISCUSSION TITLE
Native GTK Titlebar seems Broken in 1.0.1

### DISCUSSION THREAD

**Discussion by andyl** [ghostty-org/ghostty#4224] [Q&A] [answered]
In 1.0.0, I set `gtk-titlebar = false` to make Ghostty use the native titlebar.  Great!

In 1.0.1, I can find settings to enable the native titlebar.  Sad.

- "gtk-titlebar = false" - there is no titlebar
- "gtk-titlebar = true / window-decoration = false" - there is no titlebar
- "gtk-titlebar = true / window-decoration = true" - use the 'ghostty toolbar'

Is there a way to display the native titlebar in 1.0.1??

**Comment by rhodes-b** [comment, marked as answer]
does `gtk-adwaita = false` display the native titlebar? (im on a gtk based system so I cant really tell but I know it does effect it)

**Reply by rhodes-b** [reply]
@jcollie @tristan957 want to confirm this is expected behavior

**Reply by ghost** [reply]
> @jcollie @tristan957 want to confirm this is expected behavior

If this is expected behavior then pls document the relationship between 'gtk-adwaita = false' and 'window_decoration = true' in https://ghostty.org/docs/config/reference#window-decoration.

**Comment by andyl** [comment]
@rhodes-b 

Thanks for the tip! Working now.  Here are the config settings I'm using: 
```
window-decoration = true
gtk-titlebar      = false
gtk-adwaita       = false
```

**Comment by jcollie** [comment]
A lot of the problem here is that the interaction between `gtk-titlebar`, `gtk-adwaita`, `window-decoration`, and `toggle_window_decoration` is poorly specified. We should use this discussion to more fully specify the interaction.

My questions:

1. It would seem to me that if someone set `gtk-titlebar=false` that they would _never_ want the titlebar to show, regardless of toggling window decorations. On the other hand, the documentation doesn't specify this so I can see why some people would see this as an "initial state" only and toggling window decorations would show/hide the titlebar.
2. The same goes for `gtk-titlebar=true` being explicitly set in the config file.
3. If `window-decoration=false` implies that the titlebar is not shown, how would you ever get a window that has the title bar, but not the standard drop shadow and rounded corners of GTK windows?

I really think to avoid going crazy, we need to separate concerns here a bit.

My suggestion:

1. The `window-decoration` config and `toggle_window_decoration` keybind action _never_ affect the titlebar, but only the rounded corners and drop shadow that GTK adds to window (or any other bling GTK might add).
2. `gtk-titlebar` only affects the _initial_ state of the titlebar when creating a window (right now if `gtk-titlebar=false` the titlebar is never even created and can't be toggled).
3. We introduce a new keybind action `toggle_titlebar` that will hide/show the titlebar when called.
4. We introduce a new keybind action `toggle_window_decorations_and_titlebar` that will hide/show the titlebar _and_ window decorations when called. Of course you'd need to be careful using all three toggles in combination.
5. `gtk-adwaita` only controls the _style_ of the titlebar (Adwaita vs plain GTK).

I think that should be able to support any (reasonable) config and not drive us crazy trying to implement. Also note that being able to load custom CSS code is in the works so eventually you'll be able to really customize the window to your liking.

**Reply by rhodes-b** [reply]
## Questions
> 1. It would seem to me that if someone set gtk-titlebar=false that they would never want the titlebar to show, regardless of toggling window decorations. On the other hand, the documentation doesn't specify this so I can see why some people would see this as an "initial state" only and toggling window decorations would show/hide the titlebar.  

This makes sense to me, You explicitly said I dont want a titlebar so there should be no way to ever get one.

> 2. The same goes for `gtk-titlebar=true` being explicitly set in the config file.

Ditto

> 3. If `window-decoration=false` implies that the titlebar is not shown, how would you ever get a window that has the title bar, but not the standard drop shadow and rounded corners of GTK windows?

Window decoration only affecting `decoration` and not functionality like a titlebar sounds good to me
 
## Suggestion
> 1. The `window-decoration` config and `toggle_window_decoration` keybind action _never_ affect the titlebar, but only the rounded corners and drop shadow that GTK adds to window (or any other bling GTK might add).

:+1: 

>  2. `gtk-titlebar` only affects the _initial_ state of the titlebar when creating a window (right now if `gtk-titlebar=false` the titlebar is never even created and can't be toggled).

:+1: 

>     3. We introduce a new keybind action `toggle_titlebar` that will hide/show the titlebar when called.

Alot of people have asked for a feature like this already, "How to disable titlebar only when I go into fullscreen" I think the generic action covers that case nicely 

>     4. We introduce a new keybind action `toggle_window_decorations_and_titlebar` that will hide/show the titlebar _and_ window decorations when called. Of course you'd need to be careful using all three toggles in combination.

I feel like this one / adding too many runs the risk of edge cases. Like you stated how youd need to be careful

>     5. `gtk-adwaita` only controls the _style_ of the titlebar (Adwaita vs plain GTK).

I could see this going either way especially with the custom css route. If Ghostty wants to be native to Gnome it seems like adwaita should be leaned on more. If Ghostty wants to also feel native in other GTK based DE (Im a Cinnamon user myself) then it should do less like suggested here

**Reply by jcollie** [reply]
> > ```
> > 5. `gtk-adwaita` only controls the _style_ of the titlebar (Adwaita vs plain GTK).
> > ```
> 
> I could see this going either way especially with the custom css route. If Ghostty wants to be native to Gnome it seems like adwaita should be leaned on more. If Ghostty wants to also feel native in other GTK based DE (Im a Cinnamon user myself) then it should do less like suggested here

The `gtk-adwaita` option goes deeper than just appearance, it runs very different code under the hood (`GtkApplicationWindow` vs `AdwApplicationWindow` and `GtkHeaderBar` vs `AdwHeaderBar`).

There are a few developers that want to eliminate the ability to compile/run without Adwaita as it _really_ complicates the code, but we're not there yet.

**Reply by jcollie** [reply]
Actually, I like something like this much better:
```
keybind = alt+x=toggle:decoration,titlebar
```
Keeps it simple and easy to extend in the future.

**Reply by jcollie** [reply]
> > ```
> >  We introduce a new keybind action `toggle_titlebar` that will hide/show the titlebar when called.
> > ```
> 
> Alot of people have asked for a feature like this already, "How to disable titlebar only when I go into fullscreen" I think the generic action covers that case nicely

Actually, I kinda think that the titlebar should be hidden automatically when going into fullscreen, but perhaps there should be a config to enable/disable that behavior.

**Comment by ghost** [comment]
Just to clarify: This is the behavior some users want, or want to keep: https://github.com/ghostty-org/ghostty/discussions/4324
As long as it is possible to opt-in the native VM decoration title-bar, and at the same time opt-out the GTK one, many users will be happy.
That is the only request here.

**Comment by Gelmo** [comment]
Related discussion - https://github.com/ghostty-org/ghostty/issues/2023

**Comment by arodland** [comment]
I use `gtk-titlebar=false` because I want "normal" window decorations, not ghostty's big tall titlebar.

With `gtk-adwaita=false`, that's what happens. On 1.0, that's also what happened with adwaita turned on. But with current main, setting `gtk-titlebar=false` with adwaita enabled results in no titlebar at all.

And yes, it's understandable that "no titlebar" means "no titlebar" (even if that's not what I *want*), but it's entirely weird that "no titlebar" means two different things depending on `gtk-adwaita`.

**Reply by tristan957** [reply]
This sounds like a bug, and I will investigate. Thank you for the report :)

**Reply by jcollie** [reply]
I think that there's something in vanilla GTK that tells the DE to enable server-side decorations if a `GtkApplicationWindow` does not have a titlebar.  I haven't dug into the details to confirm that but our hope is to make the Ghostty experience as smooth as possible on non-Gnome DEs.

**Reply by tristan957** [reply]
I did some testing.

```c
#include <adwaita.h>
#include <gtk/gtk.h>


static void
activate (GtkApplication* app,
          gpointer        user_data)
{
  GtkWidget *window;

  window = adw_application_window_new (app);
  gtk_window_set_title (GTK_WINDOW (window), "Window");
  gtk_window_set_default_size (GTK_WINDOW (window), 200, 200);
  gtk_window_present (GTK_WINDOW (window));
}

int
main (int    argc,
      char **argv)
{
  AdwApplication *app;
  int status;

  app = adw_application_new ("org.gtk.example", G_APPLICATION_DEFAULT_FLAGS);
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);
  status = g_application_run (G_APPLICATION (app), argc, argv);
  g_object_unref (app);

  return status;
}
```

This also has no titlebar.

**Reply by tristan957** [reply]
I'm not really tempted to fix this. If you want SSDs, you should probably not be using Adwaita. Nevertheless I've asked in `#gtk:matrix.org` about this because I'm curious.

**Reply by tristan957** [reply]
![image](https://github.com/user-attachments/assets/b030fd5a-c02c-416d-92b6-0e7ee8a156fe)

> The AdwWindow widget is a subclass of [GtkWindow](https://docs.gtk.org/gtk4/class.Window.html) which has no titlebar area.

**Reply by arodland** [reply]
Sure. Personally I'm fine with turning adwaita off to get where I want to be.

I'm just thinking about clarity and consistency of options: can a user find out what choices are available to them, can they figure out how to get the behavior they want, and once they've found it will it keep working or will the next patch release change things around again?

**Reply by rhodes-b** [reply]
> Sure. Personally I'm fine with turning adwaita off to get where I want to be.
> 
> I'm just thinking about clarity and consistency of options: can a user find out what choices are available to them, can they figure out how to get the behavior they want, and once they've found it will it keep working or will the next patch release change things around again?


The point of what we are talking about is so there is a consistent behavior that doesnt change between patches. Like also stated the behavior of 1.0.0 and 1.0.1 are not correct. Which is why it may change again but we want to make it consistent going forward after that

**Reply by gtjoseph** [reply]
I didn't see this discussion when I created #5127 yesterday.  Is there a final decision/direction on getting the native titlebar back?  Right now no combination of options seems to do the trick.

**Reply by arodland** [reply]
@gtjoseph as of HEAD it seems to work with `window-decoration = server` (with or without adwaita). That's a new value for that option that isn't available in 1.0.1.

**Reply by arodland** [reply]
#4724

**Reply by gtjoseph** [reply]
@arodland Not working for me.  The commit does say "wayland" and I'm on Fedora 41 Xfce so no wayland for me.

**Comment by robertgzr** [comment]
i noticed another detail with the recent decorations changes. https://github.com/ghostty-org/ghostty/blob/72d085525b22d66468c5969a4d507a0fa68d4a04/src/apprt/gtk/winproto/wayland.zig?plain=1#L255

this makes it so that when using a compositor ([river](https://codeberg.org/river/river) for example) that doesn't implement the [Kwin decoration manager](https://wayland.app/protocols/kde-server-decoration) protocol, which is deprecated in favor of [XDG decoration manager](https://wayland.app/protocols/xdg-decoration-unstable-v1), ghostty will ignore my config entry requesting no decorations and give me some anyway.

i would expect it to just change whatever the default is, based on availability of the protocol?
